### PR TITLE
Remove Nested async and Fallibility from ObjectStore::list

### LINKED ISSUE #4946: Simplify ObjectStore::List
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Having two layers of asynchrony and fallibility makes for a rather obnoxious interface. Removing this makes for better client ergonomics, at the expense of slightly more arcane logic to implement the ObjectStore

**Describe the solution you'd like**

I would like `ObjectStore::list` to synchronously return an async stream

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4946.

# Rationale for this change
 

Having two layers of asynchrony and fallibility makes for a rather obnoxious interface. Removing this makes for better client ergonomics, at the expense of slightly more arcane logic to implement the ObjectStore

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/local.rs [hunk 1]
```diff
@@ -1158,21 +1156,14 @@ mod tests {
 
         let store = LocalFileSystem::new_with_prefix(root.path()).unwrap();
 
-        // `list` must fail
```
Comment (tustvold) [inline object_store/src/local.rs:1161]: We no longer have this peculiar arrangement where an error can emerge in two different places

File: object_store/src/limit.rs [hunk 2]
```diff
@@ -147,23 +147,24 @@ impl<T: ObjectStore> ObjectStore for LimitStore<T> {
...
         &self,
         prefix: Option<&Path>,
         offset: &Path,
-    ) -> Result<BoxStream<'_, Result<ObjectMeta>>> {
-        let permit = Arc::clone(&self.semaphore).acquire_owned().await.unwrap();
-        let s = self.inner.list_with_offset(prefix, offset).await?;
-        Ok(PermitWrapper::new(s, permit).boxed())
+    ) -> BoxStream<'_, Result<ObjectMeta>> {
+        let s = self.inner.list_with_offset(prefix, offset);
+        let fut = Arc::clone(&self.semaphore)
```
Comment (tustvold) [inline object_store/src/limit.rs:164]: This somewhat peculiar ordering is required because the returned stream cannot borrow prefix

**Review by wjones127 — COMMENTED** [review]
@tustvold this seems reasonable. I'd be +1 on this change.

**Review by alamb — COMMENTED** [review]
I agree that this is a nicer interface than an async function that returns a stream

**Review by wjones127 — APPROVED** [review]
Looks good. Should we have an issue for this?

**Comment by alamb** [comment]
cc @carols10cents 

> Looks good. Should we have an issue for this?

Looks like @tustvold  filed https://github.com/apache/arrow-rs/issues/4946

**Comment by alamb** [comment]
I think this means we need to release another SemVer change for object_store on the next release

**Comment by tustvold** [comment]
> I think this means we need to release another SemVer change for object_store on the next release

Indeed, there are already also some other breaking changes already merged and some others in the pipeline

**Comment by alamb** [comment]
I wonder if it was intended that `list_with_delimiter` is still async and returns `Result`?

It seems a strangely different API 

https://docs.rs/object_store/latest/object_store/trait.ObjectStore.html#tymethod.list_with_delimiter

**Comment by tustvold** [comment]
> I wonder if it was intended that list_with_delimiter is still async and returns Result?

Yes it was intended as list_with_delimiter does not return a stream

**Comment by alamb** [comment]
BTW I think this change has also made it somewhat harder to implement "fail fast" semantics for object store clients, something we saw in IOx and @andygrove  noticed while updating ballista

**Comment by andygrove** [comment]
I ended up with the following to resolve this:

```rust
stream::once(async { Err(...)  }).boxed()
```

**Pull request merged**